### PR TITLE
Rewrite of the getDelegationFee endpoint

### PR DESCRIPTION
**Pull request by Unisay**
Rewrite of the `getDelegationFee` endpoint so that:
1.  it uses balanceTransaction as its primary means of transaction balancing;
2.  it no longer directly depends upon the internal coin selection library.
 
### Issue Number

ADP-2264

### CODE REVIEW

**Comment by Unisay** [comment]
bors try

File: lib/wallet/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -2806,100 +2811,133 @@ migrationPlanToSelectionWithdrawals plan rewardWithdrawal outputAddressesToCycle
...
+-- In the returned 'FeeSpread', the minimum fee is that which 90% of the sampled
+-- fees are greater than. The maximum fee is the highest fee observed in the
+-- samples.
+calculateFeeSpread
     :: forall m. Monad m
-    => ExceptT ErrSelectAssets m Coin
-    -> ExceptT ErrSelectAssets m FeeEstimation
-estimateFee
+    => ExceptT ErrSelectAssets m Fee
+    -> ExceptT ErrSelectAssets m FeeSpread
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet.hs]: Ah, I take this `Fee -> FeeSpread/Estimation` is why you renamed it, but I'm not convinced as:
1. "FeeSpread" sounds like the _width_ of the distribution of fees, whereas we're also interested in the absolute values.
2. Why add the cognitive overhead of renaming functions we 1) want to remove when we are able to 2) now differ in naming compared to the endpoint handlers and documentation, which still use the "estimate fee" terminology? I don't think this makes sense.

File: lib/wallet/src/Cardano/Wallet.hs [hunk 2]
```diff
@@ -2806,100 +2811,133 @@ migrationPlanToSelectionWithdrawals plan rewardWithdrawal outputAddressesToCycle
...
-        Left e -> throwIO $ ExceptionNoSuchWallet e
+    throwInIO :: forall a. ExceptT ErrNoSuchWallet IO a -> IO a
+    throwInIO = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
 
-    db = ctx ^. dbLayer @IO @s @k
-    nl = ctx ^. networkLayer
+data DelegationFee = DelegationFee
+    { feeSpread :: FeeSpread
+    , deposit :: Coin
+    }
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet.hs]: 👍

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 3]
```diff
@@ -3473,46 +3472,32 @@ joinStakePool
...
+    AnyRecentEra (recentEra :: WriteTx.RecentEra era) <- guardIsRecentEra era
+    withWorkerCtx ctx walletId liftE liftE $ \workerCtx -> liftHandler $ do
+        W.DelegationFee {feeSpread, deposit} <-
+            W.delegationFee
+                (workerCtx ^. dbLayer)
+                netLayer
+                txLayer
+                (timeInterpreter netLayer)
+                (AnyRecentEra recentEra)
+                (W.defaultChangeAddressGen (delegationAddress @n))
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:3498]: Ah, right we're working with Shelley wallets here, there's no need for dummy addresses 👍

File: lib/wallet/src/Cardano/Wallet.hs (hunk 2 above)
Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet.hs]: (Though — is it not weird to call the inner thing Fee**Spread** and the outer thing **Fee**?)

File: lib/wallet/src/Cardano/Wallet.hs (hunk 1 above)
Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet.hs]: Possibly `FeeDistribution` or something similar is better?

**Review by Anviking — COMMENTED** [review]
LGTM, but not sold on the `FeeSpread` terminology (see comment)

File: lib/wallet/src/Cardano/Wallet.hs (hunk 1 above)
Comment (Unisay) [reply lib/wallet/src/Cardano/Wallet.hs]: Changed to `(Percentile 10 Fee, Percentile 90 Fee)`

File: lib/wallet/src/Cardano/Wallet.hs (hunk 2 above)
Comment (Unisay) [reply lib/wallet/src/Cardano/Wallet.hs]: Changed

File: lib/wallet/src/Cardano/Wallet.hs [hunk 4]
```diff
@@ -2806,100 +2813,127 @@ migrationPlanToSelectionWithdrawals plan rewardWithdrawal outputAddressesToCycle
...
+                    case recentEra of {}
+
+        deposit <- liftIO $ calcMinimumDeposit db netLayer walletId
+        pure DelegationFee { feePercentiles, deposit }
+
+-- | For a given transaction fee estimation calculation evaluates its fee spread
+-- by repeatedly running it (100 times) and recording the results.
+-- In the returned 'FeeSpread', the minimum fee is that which 90% of the sampled
+-- fees are greater than. The maximum fee is the highest fee observed in the
+-- samples.
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet.hs]: `FeeSpread` is now out of date.

And I was going to say this comment now seems redundant with the descriptive types:

> the minimum fee is that which 90% of the sampled fees are greater than. The maximum fee is the highest fee observed in the samples.

but interestingly it claims the second one is the maximum and not the 90th percentile? 😮 I can't tell immediately from the implementation — is this correct?

**Comment by Unisay** [comment]
bors try

File: lib/wallet/src/Cardano/Wallet.hs (hunk 4 above)
Comment (Unisay) [reply lib/wallet/src/Cardano/Wallet.hs]: I checked the implementation and it returns 1st and 9th decile (aka 10th and 90th percentile) values; Fixed the comment;

**Review by Anviking — APPROVED** [review]
Thanks!

**Pull request merged**